### ISSUE TITLE
Pods stuck in Pending status - kube-scheduler 1.19.13

### ISSUE THREAD

**Issue by yhrunyk** [kubernetes/kubernetes#104521]
#### What happened:
After upgrading core k8s components from 1.18 to 1.19.13 we faced a problem with Pending pods. From time to time PODs(different deployments) are stuck in Pending status. From kube-scheduler logs, we see a new event that starts appearing after the upgrade. For example `Pod kube-system/prometheus-thanos-frontend-86f4f8df77-h52tl **doesn't exist in informer cache**: pod "prometheus-thanos-frontend-86f4f8df77-h52tl" not found`.
However, ClusterAutoscaler still tries checking if there's an appropriate node `Pod kube-system.prometheus-thanos-frontend-86f4f8df77-h52tl marked as unschedulable can be scheduled on node ***.eu-west-1.compute.internal (based on hinting). Ignoring in scale up.`

It looks like kube-scheduler just forgot about the pod, because after a couple of attempts it stop trying to schedule the pod and we see no logs/events for hours. On the other hand, CA still tries to schedule it even spin up a new nodes.

#### What you expected to happen:

Pending pods are assigned to nodes.

#### How to reproduce it (as minimally and precisely as possible):

We don't see any pattern. It happens time to time for a completely different deployments.

#### Anything else we need to know?:

If we restart kube-scheduler so the leader is changed the Pending Pods are being successfully scheduled.

#### Environment:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:59:43Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.13", GitCommit:"53c7b65d4531a749cd3a7004c5212d23daa044a9", GitTreeState:"clean", BuildDate:"2021-07-15T20:53:19Z", GoVersion:"go1.15.14", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): CentOS Linux 7 (Core)
- Kernel (e.g. `uname -a`): Linux ***.eu-west-1.compute.internal 3.10.0-1160.25.1.el7.x86_64 #1 SMP Wed Apr 28 21:49:45 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug): 
- Others: ClusterAutoscaler 1.19.1, kube-scheduler 1.19.13

**Comment by YuraHrunyk** [comment]
@ahg-g Yes, those pods exist because if we restart kube-scheduler(change the leader) the new leader is probably checking the status of all of pods, and the Pending pods that might be in this status for 1-2 days get scheduled to suitable nodes.

Unfortunately, there are no specific repro steps, we just run a lot of deployments, most of them get scheduled without any issues(99%), but a few are stuck in Pending status until we either restart the Pods or change kube-scheduler leader.

Regarding  "doesn't exist in informer cache", we haven't seen it before on the previous k8s version (1.18) it appears right after we upgrade to 1.19 and this issue with the Pending pods starts happening.

**Comment by matthias50** [comment]
@ahg-g, I work on the same team as @YuraHrunyk. I just wanted to clarify a bit the work-around we use to recover from this situation. All we do is kill the kube-scheduler process which is currently the leader. Then, one of the other redundant schedulers acquires the leader  lease and the pod gets scheduled right away.

We have thousands of pods getting scheduled across hundreds of nodes and this happens only very occasionally. We aren't quite sure how we could reproduce this because there isn't a clear pattern. Any guidance you have there would be appreciated.

**Comment by ahg-g** [comment]
This is interesting, the issue is that the scheduler thinks that the pod doesn't exist anymore because it can't find it in the informer cache, and so it is ignoring it completely. 

In https://github.com/kubernetes/kubernetes/pull/90660, we removed some retry logic that perhaps mask the issue you are facing. 

@liggitt If have a couple of questions:
1) If a pod used to exist in the informer cache, is there any reason that it would *temporarily* cease to exist?
2) Is it possible to have a race where we receive a pod add event, but the pod may not be readily accessible the pod lister (the event and lister are both setup from the same pod informer)?

@matthias50 can you tell us a bit more about your setup? are deploying k8s manually? do you have the scheduler running on the same VM as the api-server?

/cc @Huang-Wei @alculquicondor

**Comment by liggitt** [comment]
> If a pod used to exist in the informer cache, is there any reason that it would temporarily cease to exist?

Not that I know of. The client state (the store used by the pod lister) handled updated and relists by replacing existing content, it doesn't remove and re-add objects (_and_ it do so under lock)

> Is it possible to have a race where we receive a pod add event, but the pod may not be readily be accessible via the pod lister (the event handler and lister are both setup from the same pod informer)?

Not that I know of... the store is updated before the event handler is called

https://github.com/kubernetes/kubernetes/blob/34fb61beba52d9e1ab3f044a7be8b9f49d99534e/staging/src/k8s.io/client-go/tools/cache/controller.go#L399-L402

https://github.com/kubernetes/kubernetes/blob/34fb61beba52d9e1ab3f044a7be8b9f49d99534e/staging/src/k8s.io/client-go/tools/cache/controller.go#L404-L407

https://github.com/kubernetes/kubernetes/blob/34fb61beba52d9e1ab3f044a7be8b9f49d99534e/staging/src/k8s.io/client-go/tools/cache/controller.go#L410-L413

It is possible for the event handler to encounter a store that has already processed the *next* event received from the API server, so the following is possible:
1. pod is updated
2. pod is deleted
3. scheduler observes update event
4. scheduler updates pod in store
5. scheduler enqueues call to OnUpdate
6. scheduler observes delete event
7. scheduler removes pod from store
8. scheduler enqueues call to OnDelete
9. OnUpdate is called, looks in store, cannot find pod
10. OnDelete is called, looks in store, cannot find pod

**Comment by Huang-Wei** [comment]
The error message was logged here:

https://github.com/kubernetes/kubernetes/blob/34fb61beba52d9e1ab3f044a7be8b9f49d99534e/pkg/scheduler/factory.go#L400-L405

and it was introduced in 1.19.

Regarding the root cause, one thing to note is: the podLister used here is "filtered" (`status.phase!=Succeeded,status.phase!=Failed`) - all the way from scheduler's instantiation:

https://github.com/kubernetes/kubernetes/blob/34fb61beba52d9e1ab3f044a7be8b9f49d99534e/pkg/scheduler/scheduler.go#L756-L763

So is it possible that the pod somehow falls outside the indexer (i.e., `status.phase==Succeeded` or `status.phase==Failed`), so scheduler cannot get the desired pod?

**Comment by Huang-Wei** [comment]
@YuraHrunyk could you verify what's the eventual `status.phase` of the pending pods, and if by any chance, the _interim_ `status.phase` when the "xxx doesn't exist in informer cache" log was firstly spotted?

BTW: in 1.18, the pod was obtained from [a live API query](https://github.com/kubernetes/kubernetes/blob/1f3e19b7beb1cc0110255668c4238ed63dadb7ad/pkg/scheduler/factory.go#L498) instead of "filtered" podLister.

**Comment by ahg-g** [comment]
@Huang-Wei notice they mentioned that the pod did get eventually scheduled when switching the leader.

@YuraHrunyk  can you also please verify that the same pod, not a replacement one, was eventually scheduled?

**Comment by liggitt** [comment]
is it the informer cache / lister the pods are missing from, or the caches maintained by these:

https://github.com/kubernetes/kubernetes/blob/ebb9fafac620dac6e632b7b2607d575cc1df926f/pkg/scheduler/eventhandlers.go#L280-L329

if so, is it possible the filtering is dropping events needed to maintain those caches accurately, or could something like  https://github.com/kubernetes/kubernetes/blob/ebb9fafac620dac6e632b7b2607d575cc1df926f/pkg/scheduler/eventhandlers.go#L216-L217 explain why the pod is seen as missing elsewhere?

**Comment by ahg-g** [comment]
It is not missing from the scheduler's maintained cache, it is missing from the pod informer cache, here is the failed lookup: https://github.com/kubernetes/kubernetes/blob/44a3aa90a777aa05e997cb827744597c06b70803/pkg/scheduler/factory.go#L469) 

this pod informer is configured to exclude pods that ran to completion: https://github.com/kubernetes/kubernetes/blob/44a3aa90a777aa05e997cb827744597c06b70803/pkg/scheduler/factory.go#L432

**Comment by alculquicondor** [comment]
@YuraHrunyk can you provide all the logs referring to the pod that "disappears"? I would like to know if we tried to schedule it at other points. That might help us narrow down the possible race conditions.

**Comment by alculquicondor** [comment]
Logs from the apiserver would help too.

**Comment by matthias50** [comment]
>  can you also please verify that the same pod, not a replacement one, was eventually scheduled?

@ahg-g @YuraHrunyk I can verify that it is the same pod, not a replacement one.

**Comment by matthias50** [comment]
> can you tell us a bit more about your setup? are deploying k8s manually? do you have the scheduler running on the same VM as the api-server?

@ahg-g, we are deploying k8s using home grown automation on AWS. All our clusters have three master nodes (AWS instances), each with static pods for the k8s control plane components (kube-scheduler, kube-apiserver, kube-controller-manager).

**Comment by YuraHrunyk** [comment]
@alculquicondor Adding k8s logs.
`               !!!!!!!KUBE-SCHEDULER !!!

   log: W0819 07:44:24.714335       1 factory.go:471] Pod here-walle-lane-group-storage-sit/sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl doesn't exist in informer cache: pod "sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl" not found
   log_time: 2021-08-19T07:44:24.714347580+00:00

   log: I0819 07:44:24.714299       1 factory.go:449] "Unable to schedule pod; no fit; waiting" pod="here-walle-lane-group-storage-sit/sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl" err="0/315 nodes are available: 10 node(s) had taint {NUC: }, that the pod didn't tolerate, 10 node(s) had taint {workload: walle-default}, that the pod didn't tolerate, 13 node(s) had taint {workload: nvidia-gpu}, that the pod didn't tolerate, 2 node(s) had taint {workload: sdaa}, that the pod didn't tolerate, 259 Insufficient cpu, 3 node(s) had taint {workload: prometheus}, that the pod didn't tolerate, 4 node(s) were unschedulable, 6 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate, 6 node(s) had taint {workload: opensearch}, that the pod didn't tolerate, 9 Insufficient memory."
   log_time: 2021-08-19T07:44:24.714347580+00:00

   log: W0819 07:44:24.693030       1 factory.go:471] Pod here-walle-lane-group-storage-sit/sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl doesn't exist in informer cache: pod "sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl" not found
   log_time: 2021-08-19T07:44:24.693069991+00:00

   log: I0819 07:44:24.692971       1 factory.go:449] "Unable to schedule pod; no fit; waiting" pod="here-walle-lane-group-storage-sit/sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl" err="0/315 nodes are available: 10 node(s) had taint {NUC: }, that the pod didn't tolerate, 10 node(s) had taint {workload: walle-default}, that the pod didn't tolerate, 13 node(s) had taint {workload: nvidia-gpu}, that the pod didn't tolerate, 2 node(s) had taint {workload: sdaa}, that the pod didn't tolerate, 259 Insufficient cpu, 3 node(s) had taint {workload: prometheus}, that the pod didn't tolerate, 4 node(s) were unschedulable, 6 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate, 6 node(s) had taint {workload: opensearch}, that the pod didn't tolerate, 9 Insufficient memory."
   log_time: 2021-08-19T07:44:24.693034560+00:00

               !!!!!!! KUBE-APISERVER  !!!!!

   log: 2021-08-19T07:44:24.709694908Z AUDIT: id="32212327-1493-4b57-91b6-ff3c5ba53091" stage="ResponseComplete" ip="***" method="get" user="system:kube-scheduler" groups="\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="here-walle-lane-group-storage-sit" uri="/api/v1/namespaces/here-walle-lane-group-storage-sit/pods/sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl" response="200"
   log_time: 2021-08-19T07:44:24.712727788+00:00

   log: 2021-08-19T07:44:24.694084101Z AUDIT: id="e1bb9e70-4f48-47bc-ade9-9c4b38dfbbd3" stage="ResponseComplete" ip="***" method="patch" user="system:kube-scheduler" groups="\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="here-walle-lane-group-storage-sit" uri="/api/v1/namespaces/here-walle-lane-group-storage-sit/pods/sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl/status" response="200"
   log_time: 2021-08-19T07:44:24.706216120+00:00

   log: 2021-08-19T07:44:24.684012662Z AUDIT: id="3f7a37d8-f6b1-4d1c-858b-7cb31b54cbff" stage="ResponseComplete" ip="***" method="get" user="system:kube-scheduler" groups="\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="here-walle-lane-group-storage-sit" uri="/api/v1/namespaces/here-walle-lane-group-storage-sit/pods/sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl" response="200"
   log_time: 2021-08-19T07:44:24.691003190+00:00

            !!!!!!!!! Cluster Autoscaler !!!!!

   log: I0820 14:51:22.260117       1 cluster.go:306] Pod here-walle-lane-group-storage-sit/sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl can be moved to ***.eu-west-1.compute.internal
   log_time: 2021-08-20T14:51:22.260173526+00:00

   log: I0820 14:51:21.958325       1 filter_out_schedulable.go:118] Pod here-walle-lane-group-storage-sit.sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl marked as unschedulable can be scheduled on node ***.eu-west-1.compute.internal (based on hinting). Ignoring in scale up.
   log_time: 2021-08-20T14:51:21.958367479+00:00

            !!!!!! KUBE-CONTROLLER-MANAGER !!!!

   log: I0819 07:44:24.758373       1 event.go:291] "Event occurred" object="here-walle-lane-group-storage-sit/sit2-walle-lane-group-storage-event-publisher-deployment-864c9d9b8b" kind="ReplicaSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulCreate" message="Created pod: sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl"
   log_time: 2021-08-19T07:44:24.758464495+00:00

            !!!!!!!! OTHER Events!!!

   involvedObjectKind: ReplicaSet
   message: Created pod: sit2-walle-lane-group-storage-event-publisher-deployment-8r5rkl
   namespace: here-walle-lane-group-storage-sit

   involvedObjectKind: Pod
   k8scluster: ***-aws-eu-west-1
   message: pod triggered scale-up: [{5.default.nodes.***-aws-eu-west-1.k8s.in.here.com 19->20 (max: 130)}]`

**Comment by alculquicondor** [comment]
`07:44:24.758464495`: we have a confirmation event that the pod was just created.

`07:44:24.693034560`: we see that the scheduler already tried to schedule it at and failed.

`07:44:24.693069991`: The scheduler fails to see the pod in the informer cache.

So it seems to be the case that the scheduler got the "Add Pod" event, but the pod was not added to the informer cache.

Unfortunately, 1.19 is not a supported version anymore. Could you try to see if the issue persist in 1.20?

**Comment by matthias50** [comment]
@alculquicondor, https://kubernetes.io/releases/ says the end of life for 1.19 is 2021-10-28. I get that we are close to that date, but we haven't passed it yet. So it seems like 1.19 should still be supported.

**Comment by alculquicondor** [comment]
Oops, sorry about that. I was going by the high level rule of thumb: 3 latest releases are supported.

Maybe this could be reproduced at integration level. Anyone willing to setup a test?

**Comment by ahg-g** [comment]
> but the pod was not added to the informer cache.

and there is reason to think it was transitory because in 1.18 we had a retry loop that was removed in https://github.com/kubernetes/kubernetes/pull/90660 that likely masked this failed lookup

**Comment by yhrunyk** [comment]
@alculquicondor @ahg-g I enabled v4 logging for the kube-scheduler. Not much more logs, though you might find it useful.

`
   log: I0913 13:13:24.391670       1 eventhandlers.go:173] add event for unscheduled pod example-deploy/example-pod
   log_time: 2021-09-13T13:13:24.391795782+00:00

   log: I0913 13:13:24.391912       1 eventhandlers.go:173] add event for unscheduled pod example-deploy/example-pod
   log_time: 2021-09-13T13:13:24.392002904+00:00
  
   log: I0913 13:13:24.391989       1 scheduling_queue.go:832] About to try and schedule pod example-deploy/example-pod
   log_time: 2021-09-13T13:13:24.392002904+00:00
   
   log: I0913 13:13:24.392002       1 scheduler.go:464] Attempting to schedule pod: example-deploy/example-pod
   log_time: 2021-09-13T13:13:24.392083105+00:00
   
   log: I0913 13:13:24.408202       1 eventhandlers.go:173] add event for unscheduled pod example-deploy/example-pod
   log_time: 2021-09-13T13:13:24.408782995+00:00
   
   log: I0913 13:13:24.436456       1 factory.go:449] "Unable to schedule pod; no fit; waiting" pod="example-deploy/example-pod" err="0/323 nodes are available: 17 Insufficient memory, 2 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate, 278 Insufficient cpu, 42 node(s) didn't match node selector."
   log_time: 2021-09-13T13:13:24.436529751+00:00
   
   log: W0913 13:13:24.436525       1 factory.go:471] Pod example-deploy/example-pod doesn't exist in informer cache: pod "example-pod" not found
   log_time: 2021-09-13T13:13:24.436585872+00:00
   
   log: I0913 13:13:24.436851       1 scheduler.go:359] Updating pod condition for example-deploy/example-pod to (PodScheduled==False, Reason=Unschedulable)
   log_time: 2021-09-13T13:13:24.437692798+00:00
   
   log: I0913 13:13:24.479475       1 scheduling_queue.go:832] About to try and schedule pod example-deploy/example-pod
   log_time: 2021-09-13T13:13:24.479528587+00:00
  
   log: I0913 13:13:24.479513       1 scheduler.go:464] Attempting to schedule pod: example-deploy/example-pod
   log_time: 2021-09-13T13:13:24.479528587+00:00
   
   log: I0913 13:13:24.492416       1 factory.go:449] "Unable to schedule pod; no fit; waiting" pod="example-deploy/example-pod" err="0/323 nodes are available: 17 Insufficient memory, 2 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate, 278 Insufficient cpu, 42 node(s) didn't match node selector."
   log_time: 2021-09-13T13:13:24.492487365+00:00
   
   log: W0913 13:13:24.492459       1 factory.go:471] Pod example-deploy/example-pod doesn't exist in informer cache: pod "example-pod" not found
   log_time: 2021-09-13T13:13:24.492487365+00:00
 
   log: I0913 13:13:24.492540       1 scheduler.go:359] Updating pod condition for example-deploy/example-pod to (PodScheduled==False, Reason=Unschedulable)
   log_time: 2021-09-13T13:13:24.492579987+00:00


               !!!!! After the kube-scheduler leader was changed !!!

   
   log: I0913 13:24:08.562411       1 eventhandlers.go:173] add event for unscheduled pod example-deploy/example-pod
   log_time: 2021-09-13T13:24:08.562882816+00:00
   
   log: I0913 13:24:21.109330       1 scheduling_queue.go:832] About to try and schedule pod example-deploy/example-pod
   log_time: 2021-09-13T13:24:21.109354308+00:00
   
   log: I0913 13:24:21.109346       1 scheduler.go:464] Attempting to schedule pod: example-deploy/example-pod
   log_time: 2021-09-13T13:24:21.109354308+00:00
   
   log: I0913 13:24:21.131437       1 default_binder.go:51] Attempting to bind example-deploy/example-pod to ip-10-104-177-147.eu-west-1.compute.internal
   log_time: 2021-09-13T13:24:21.131458262+00:00
   
   log: I0913 13:24:21.139439       1 scheduler.go:609] "Successfully bound pod to node" pod="example-deploy/example-pod" node="***.eu-west-1.compute.internal" evaluatedNodes=327 feasibleNodes=3
   log_time: 2021-09-13T13:24:21.139541417+00:00
   
   log: I0913 13:24:21.139810       1 eventhandlers.go:205] delete event for unscheduled pod example-deploy/example-pod
   log_time: 2021-09-13T13:24:21.139892032+00:00
   
   log: I0913 13:24:21.139839       1 eventhandlers.go:225] add event for scheduled pod example-deploy/example-pod 
   log_time: 2021-09-13T13:24:21.139892032+00:00
   
   log: I0913 13:41:40.172389       1 eventhandlers.go:283] delete event for scheduled pod example-deploy/example-pod 
   log_time: 2021-09-13T13:41:40.172457829+00:00
   `

**Comment by alculquicondor** [comment]
Could you confirm if these are logs from all the schedulers running (even the ones that are not leaders)? Just trying to make sure if the three log lines for `add event for unscheduled pod example-deploy/example-pod` are expected.

**Comment by alculquicondor** [comment]
Given what @liggitt said in https://github.com/kubernetes/kubernetes/issues/104521#issuecomment-914657017 -- discarding ordering issues in the client code -- another possibility is that we are creating more than one informer factory, one which we use for the lister and the other where we register the events. I'm looking into this, but so far I haven't found any problems.

**Comment by yhrunyk** [comment]
@alculquicondor Yes, those logs come from all of the kube-scheduler replicas, we have 3 of them.

**Comment by alculquicondor** [comment]
I was right: We had 2 informers in 1.19. We fixed that in 1.20 via #93044
Let me see if I can cherry pick.

**Comment by Huang-Wei** [comment]
> I was right: We had 2 informers in 1.19. We fixed that in 1.20 via #93044

That sounds promising. Thanks for digging into the problem!

BTW: it'd be great if @yhrunyk can confirm the problem is gone in 1.20.

**Comment by alculquicondor** [comment]
FYI, we are a bit tight on the deadline for 1.19 patch releases. I'm asking sig-release if we can squeeze this cherry-pick.

**Comment by ahg-g** [comment]
> I was right: We had 2 informers in 1.19. We fixed that in 1.20 via #93044

so I looked at that before, but couldn't find places where the InformerFactory pod informer is used. The separately created pod informer was used to setup the events and lookup the pod where the error was triggered.

**Comment by alculquicondor** [comment]
This is the PR that introduced the lister check, and you can see that it used the shared informer factory lister https://github.com/kubernetes/kubernetes/commit/133a02540fce435996a1104357e123b9f5b29093#diff-53cc60f8818dce859b9222be5c8dc93d511fbc942cc17a0e85c10f260d74ca72R201

**Comment by Huang-Wei** [comment]
Here is a summary on the underlying problem across releases:

- in k8s versions <= 1.19, scheduler maintains two podInformers - one is from the sharedInformerFactory the other is created independently (code of v1.19.14 below). This is problematic.
    https://github.com/kubernetes/kubernetes/blob/0fd2b5afdfe3134d6e1531365fdb37dd11f54d1c/cmd/kube-scheduler/app/options/options.go#L288-L289
- in theory, the problem (described in this issue) should surface in all versions, but why it's only reported until 1.19? It's because in the versions <= 1.18, we call API server directly in the error handling logic (instead of podLister). And API server always has the source of truth (code of v1.18.20 below)
    https://github.com/kubernetes/kubernetes/blob/1f3e19b7beb1cc0110255668c4238ed63dadb7ad/pkg/scheduler/factory.go#L498
- what happened in 1.19? You may have noticed that we removed the direct API call (as unneeded API server queries is an anti-pattern), and use the podLister instead. It's supposed to work properly, but without querying API server, the 2 podInformer must be unified; otherwise, the 2 independent informers may have racing issues, as what this issue reported.
- what happened in 1.20? We unified the 2 podInformers and thought that was a _good-to-have_ fix, but it turns out to be a must. So Aldo raised #105015 to back port the fix to 1.19.

**Comment by ahg-g** [comment]
ah, I missed this https://github.com/kubernetes/kubernetes/blob/0fd2b5afdfe3134d6e1531365fdb37dd11f54d1c/pkg/scheduler/factory.go#L202

in theory, the fix to 1.19 could have been to only change the line above to use the independent pod informer.

**Comment by alculquicondor** [comment]
> in theory, the fix to 1.19 could have been to only change the line above to use the independent pod informer.

Yeah, I considered that, but it's also theoretically more risky, as we don't have soak time for such change. See https://github.com/kubernetes/kubernetes/pull/105015#issuecomment-919574618 for context.

**Comment by ahg-g** [comment]
it would be good if the OP can verify the fix and report back.

**Comment by yhrunyk** [comment]
@ahg-g I will provide feedback as soon as we deploy it to our cluster.

**Comment by alculquicondor** [comment]
The fix is now in 1.19.15 :)

**Comment by frvge** [comment]
I'm on the same team as OP. The cherry-pick fixed it. Thank you all for the quick resolution on a very tight deadline.
